File: Frontend/src/components/services/userService.ts
// src/services/userService.ts
import axios from 'axios';

const API_BASE_URL = 'http://172.236.144.75:8080/api/user';

/**
 * User interface matching the backend User model.
 */
export interface User {
  id: number;
  username: string;
  email: string;
  password: string;
  role: string;
}

const userService = {
  /**
   * getAll
   * ------
   * Retrieves all users.
   * Endpoint: GET /users
   *
   * @returns A promise resolving to an array of User objects.
   */
  getAll: async (): Promise<User[]> => {
    const response = await axios.get(`${API_BASE_URL}/users`);
    return response.data;
  },

  /**
   * getById
   * -------
   * Retrieves a single user by their ID.
   * Endpoint: GET /users/{id}
   *
   * @param id - The user's ID.
   * @returns A promise resolving to the User object.
   */
  getById: async (id: number | string): Promise<User> => {
    const response = await axios.get(`${API_BASE_URL}/users/${id}`);
    return response.data;
  },

  /**
   * create
   * ------
   * Creates a new user.
   * Endpoint: POST /admin/users
   *
   * @param user - The user data (excluding the auto-generated id).
   * @returns A promise resolving to the newly created User object.
   */
  create: async (user: Omit<User, 'id'>): Promise<User> => {
    const response = await axios.post(`${API_BASE_URL}/admin/users`, user);
    return response.data;
  },

  /**
   * update
   * ------
   * Updates an existing user.
   * Endpoint: PUT /users/{id}
   *
   * @param id - The user's ID.
   * @param user - Partial user data to update.
   * @returns A promise resolving to the updated User object.
   */
  update: async (id: number | string, user: Partial<User>): Promise<User> => {
    const response = await axios.put(`${API_BASE_URL}/users/${id}`, user);
    return response.data;
  },

  /**
   * delete
   * ------
   * Deletes a user by their ID.
   * Endpoint: DELETE /users/{id}
   *
   * @param id - The user's ID.
   * @returns A promise that resolves when deletion is complete.
   */
  delete: async (id: number | string): Promise<void> => {
    await axios.delete(`${API_BASE_URL}/users/${id}`);
  },

  /**
   * isUser
   * ------
   * Validates user login credentials.
   * Endpoint: POST /login (adjust as needed)
   *
   * @param username - The user's username.
   * @param password - The user's raw password.
   * @returns A promise resolving to the user's role if credentials are valid;
   *          otherwise, returns "null" (as a string).
   */
  isUser: async (username: string, password: string): Promise<string> => {
    const response = await axios.post(`${API_BASE_URL}/login`, { "username": username, "password": password });
    return response.data;
  },
};

export default userService;

File: Service/SMS/src/main/java/com/qslabs/sms/service/UserService.java
package com.qslabs.sms.service;

import com.qslabs.sms.dto.AuthDTO;
import com.qslabs.sms.dto.UserDTO;
import com.qslabs.sms.dto.UserResponseDTO;
import com.qslabs.sms.model.User;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

/**
 * Service interface for managing users.
 * Handles operations like registration, authentication, and profile updates.
 */
public interface UserService {

    /**
     * Creates a new user with encoded password and stores it in the system.
     *
     * @param userDTO user registration data
     * @return ID of the newly created user
     */
    Long createUser(UserDTO userDTO);

    Page<UserResponseDTO> getAllUsers(Pageable pageable);

    /**
     * Validates user credentials.
     *
     * @param username the username entered during login
     * @param password the raw password to validate
     * @return role of the user if valid, otherwise "null" string
     */
    AuthDTO isUser(String username, String password);

    /**
     * Updates the username of a user by ID.
     *
     * @param id      user ID
     * @param userDTO object containing the new username
     * @return updated UserDTO
     */
    UserDTO updateUserName(Long id,UserDTO userDTO);

    /**
     * Updates the email address of a user by ID.
     *
     * @param id      user ID
     * @param userDTO object containing the new email
     * @return updated UserDTO
     */
    UserDTO updateEmail(Long id,UserDTO userDTO);

    /**
     * Updates the user's password (should be encoded before saving).
     *
     * @param id      user ID
     * @param userDTO object containing the new password
     * @return updated UserDTO
     */
    UserDTO updateUserPassword(Long id, UserDTO userDTO);

    /**
     * Updates the user's role (e.g., ADMIN, TEACHER, STUDENT).
     *
     * @param id      user ID
     * @param userDTO object containing the new role
     * @return updated UserDTO
     */
    UserDTO updateUserRole(Long id, UserDTO userDTO);

    /**
     * Deletes a user from the system by ID.
     *
     * @param id ID of the user to delete
     */
    void deleteUser(Long id);
}
